### PR TITLE
CellsUsed refactoring and fix

### LINKED ISSUE #1026: CellsUsed method does not consider data validation rules
**Do you want to request a *feature* or report a *bug*?**
- [x] Bug
- [ ] Feature

**Version of ClosedXML**

0.93

**What is the current behavior?**

When the data validation rules are applied to cells these cells may be not presented in the `CellsUsed()` results if they have neither data, nor styles, nor conditional formats.

**What is the expected behavior or new feature?**

Complete this.

**Did this work in previous versions of our tool?  Which versions?**

No

**Code to reproduce problem:**
```c#
        [Test]
        public void ApplyingDataValidationMakesCellNotEmpty()
        {
            var wb = new XLWorkbook();
            var ws = wb.Worksheets.Add("Sheet1");
            ws.Range("B2:B12").SetDataValidation()
                .Decimal.EqualOrGreaterThan(0);

            var usedCells = ws.CellsUsed(includeFormats: true).ToList();

            Assert.AreEqual(11, usedCells.Count);
            Assert.AreEqual("B2", usedCells.First().Address.ToString());
            Assert.AreEqual("B12", usedCells.Last().Address.ToString());
        }
```
- [ ] I attached a sample spreadsheet.  (You can drag files on to this issue)

### PR DESCRIPTION
**Pull request by Pankraty**
Fixes #1026 

I started this branch right after discovering the issue with CellsUsed method during #968 fix but then had to switch to other activities.

To do things right I started a large refactoring
* I added a flagged enum `XLCellsUsedOptions` similar to `XLClearOptions` (I'd prefer to keep them separate even though they have same values)
* I heavily refactored `CellsUsed`, `FirstCellUsed`, `LastCellUsed` and other methods to consider merged ranges, conditional formats, data validation rules, etc.
* I marked all methods using `includeFormat` flag as `[Obsolete]`. After a couple versions we can remove them I suppose.
* ~I found and fixed another bug, which perhaps I should address separately. When cells are merged they didn't keep their own styles and values instead of acquiring the style of the first cell in the range and clearing everything else (contents, data validations, conditional formats...)~ Moved to #1030

### CODE REVIEW

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 1]
```diff
@@ -1991,9 +1991,9 @@ private Boolean SetRange(Object rangeObject)
             {
                 if (!(asRange is XLRow || asRange is XLColumn))
                 {
-                    var maxRows = asRange.RowCount();
-                    var maxColumns = asRange.ColumnCount();
-                    Worksheet.Range(_rowNumber, _columnNumber, maxRows, maxColumns).Clear();
+                    var maxRow = _rowNumber + asRange.RowCount() - 1;
+                    var maxColumn = _columnNumber + asRange.ColumnCount() - 1;
+                    Worksheet.Range(_rowNumber, _columnNumber, maxRow, maxColumn).Clear();
```
Comment (igitur) [inline ClosedXML/Excel/Cells/XLCell.cs]: This was fixed in https://github.com/ClosedXML/ClosedXML/pull/1022/commits/2cc52eba90992671e4d396f3b12a7207b9f7f5fe , right?

File: ClosedXML/Excel/Cells/XLCells.cs [hunk 2]
```diff
@@ -24,15 +24,133 @@ public XLCells(bool usedCellsOnly, bool includeFormats, Func<IXLCell, Boolean> p
         {
             _usedCellsOnly = usedCellsOnly;
             _includeFormats = includeFormats;
-            _predicate = predicate;
+
+            if (predicate == null)
+                _predicate = c => true;
+            else
+                _predicate = predicate;
```
Comment (igitur) [inline ClosedXML/Excel/Cells/XLCells.cs]: `_predicate = predicate ?? _ => true;`

Comment (Pankraty) [reply ClosedXML/Excel/Cells/XLCells.cs]: Of course

File: ClosedXML/Excel/Cells/XLCell.cs (hunk 1 above)
Comment (Pankraty) [reply ClosedXML/Excel/Cells/XLCell.cs]: I noticed that too. I am going to rebase anyway, so this will be replaced with your solution.

**Comment by Pankraty** [comment]
As for the bug with the merged ranges, I found that the issue is more complicated than I thought.
If I unmerge a range in the TransposeRange.xlsx I get the following result:
image

Both the "old" way and the "new" way produce incorrect result. Guess I'll extract this into a separate issue.

**Comment by igitur** [comment]
Please see review comments in #1030

**Comment by igitur** [comment]
Had a chance to look at this further? It just needs history rewriting. I'd like release soon.

**Comment by Pankraty** [comment]
Will do it right away

**Comment by Pankraty** [comment]
I've rebased and squashed the branch and eliminated most of the issues. But more refactoring is still needed in order to properly handle large ranges in conditional formats and data validation rules (your recent test CFShiftedTruncateRange now takes forever to execute) :(

Unfortunately, I don't have enough time for this today. Will try to do it during the week.

**Review by igitur — APPROVED** [review]

**Comment by igitur** [comment]
Methods made obsolete in this PR can be removed now, right?

**Pull request merged**